api/paidAction/index.js

```diff
@@ -107,283 +202,285 @@ export default async function performPaidAction (actionType, args, context) {
+    const subContext = { ...context, tx, optimistic: false }
+    await checkUser(me?.id ?? USER_ID.anon, subContext)
+    const invoiceEntry = await createDbInvoice(invoiceData, {
+      actionType,
+      actionId,
+      optimistic: false,
+      args
+    }, subContext)
+    return {
+      invoice: invoiceEntry,
+      paymentMethod: 'PESSIMISTIC',
+      retriableWithWallet: invoiceData.retriable
+    }
   }
+  if (context.tx) return await performInvoicedAction(context.tx)
+  return await models.$transaction(performInvoicedAction, { isolationLevel: Prisma.TransactionIsolationLevel.Serializable })
+}
 
-  context.optimistic = true
-  context.me = await models.user.findUnique({ where: { id: me.id } })
-
-  const { msatsRequested, actionId } = failedInvoice
-  context.cost = BigInt(msatsRequested)
-  context.actionId = actionId
-  const invoiceArgs = await createSNInvoice(actionType, args, context)
+export async function retryPaidAction ({ invoiceId, forceInternal, resetAttempts, prioritizeInternal }, context) {
+  const { models, lnd, boss, me } = context
 
   return await models.$transaction(async tx => {
-    context.tx = tx
-
-    // update the old invoice to RETRYING, so that it's not confused with FAILED
-    await tx.invoice.update({
-      where: {
-        id: failedInvoice.id,
-        actionState: 'FAILED'
-      },
-      data: {
-        actionState: 'RETRYING'
-      }
-    })
-
-    // create a new invoice
-    const invoice = await createDbInvoice(actionType, args, context, invoiceArgs)
-
-    return {
-      result: await action.retry({ invoiceId: failedInvoice.id, newInvoiceId: invoice.id }, context),
-      invoice,
-      paymentMethod: 'OPTIMISTIC'
+    const subContext = { ...context, tx }
+    const failedInvoice = await models.invoice.findUnique({ where: { id: invoiceId, userId: me?.id ?? USER_ID.anon } })
+    if (!failedInvoice) {
+      throw new Error('invoice not found')
     }
-  }, { isolationLevel: Prisma.TransactionIsolationLevel.ReadCommitted })
-}
-
-const INVOICE_EXPIRE_SECS = 600
-const MAX_PENDING_PAID_ACTIONS_PER_USER = 100
 
-export async function assertBelowMaxPendingInvoices (context) {
-  const { models, me } = context
-  const pendingInvoices = await models.invoice.count({
-    where: {
-      userId: me?.id ?? USER_ID.anon,
-      actionState: {
-        notIn: PAID_ACTION_TERMINAL_STATES
-      }
+    if (
+      failedInvoice.actionState !== 'FAILED' &&
+      failedInvoice.actionState !== 'PENDING' &&
+      failedInvoice.actionState !== 'PENDING_HELD'
+    ) {
+      throw new Error(`actions is not in a retriable state: ${failedInvoice.actionState}`)
     }
-  })
 
-  if (pendingInvoices >= MAX_PENDING_PAID_ACTIONS_PER_USER) {
-    throw new Error('You have too many pending paid actions, cancel some or wait for them to expire')
-  }
-}
+    // cancel the previous invoice
+    await finalizeHodlInvoice({ data: { hash: failedInvoice.hash }, lnd, models, boss })
```

Comment:
Cancel the previous invoice
I am not sure if running everything in a transaction is overkill, but i suppose we don't want a race condition where the invoice being cancelled while the action is still retrying